dask/array/_array_expr/_expr.py

```diff
@@ -163,5 +189,94 @@ def unify_chunks_expr(*args):
                 changed = True
             else:
                 pass
-        arrays.extend([a, i])
+        arrays.append(a)
     return chunkss, arrays, changed
+
+
+class Stack(ArrayExpr):
+    _parameters = ["array", "axis", "meta"]
+
+    @functools.cached_property
+    def args(self):
+        return [self.array] + self.operands[len(self._parameters) :]
+
+    @functools.cached_property
+    def _meta(self):
+        return self.operand("meta")
+
+    @functools.cached_property
+    def chunks(self):
+        n = len(self.args)
+        return (
+            self.array.chunks[: self.axis]
+            + ((1,) * n,)
+            + self.array.chunks[self.axis :]
+        )
+
+    @functools.cached_property
+    def _name(self):
+        return "stack-" + _tokenize_deterministic(*self.operands)
+
+    def _layer(self) -> dict:
+        keys = list(product([self._name], *[range(len(bd)) for bd in self.chunks]))
+        names = [a.name for a in self.args]
+        axis = self.axis
+        ndim = self._meta.ndim - 1
+
+        inputs = [
+            (names[key[axis + 1]],) + key[1 : axis + 1] + key[axis + 2 :]
+            for key in keys
+        ]
+        values = [
+            (
+                getitem,
+                inp,
+                (slice(None, None, None),) * axis
+                + (None,)
+                + (slice(None, None, None),) * (ndim - axis),
+            )
+            for inp in inputs
+        ]
+        return dict(zip(keys, values))
```

Comment:
I'm not 100% about the TaskRef

Note, the keys do not necessarily be materialized at the start of the function

```suggestion
        values = [
            Task(
                key,
                getitem,
                TaskRef(inp),
                (slice(None, None, None),) * axis
                + (None,)
                + (slice(None, None, None),) * (ndim - axis),
            )
            for key, inp in zip(keys, inputs)
        ]
        return dict(zip(keys, values))
```